### PR TITLE
Add CloudStack cloud provider

### PR DESCRIPTION
**Pull request by ngtuna**
This PR aims to support Apache CloudStack cloud provider. Related issue: #26045

### CODE REVIEW

File: pkg/cloudprovider/providers/cloudstack/MAINTAINERS.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Maintainers
```
Comment (lavalamp) [inline pkg/cloudprovider/providers/cloudstack/MAINTAINERS.md]: Please add an OWNERs file with this info-- look in some of the other cloud provider packages for examples :)

**Comment by lavalamp** [comment]
ok to test

**Comment by lavalamp** [comment]
@ngtuna 1.3 feature complete date was yesterday-- is it OK with you if this goes into 1.4?

**Comment by ngtuna** [comment]
@lavalamp It's no problem to go into 1.4. Thanks :-)

**Comment by lavalamp** [comment]
:+1: Great!

**Comment by runseb** [comment]
@lavalamp that said if there is a way to get this in 1.3 , that'd be awesome ;)

**Comment by ngtuna** [comment]
@lavalamp @runseb Can you give me some hints on the failure of Jenkins unit/integration test ? I checked the log and found some clues as below:

```
Verifying ./hack/../hack/verify-godep-licenses.sh
Checking for 'Godeps/' changes against 'origin/master'
Created workspace: /go/src/k8s.io/kubernetes/_tmp/kube-godep-licenses.bPzSeq
Your godep licenses file is out of date. Run hack/update-godep-licenses.sh --create-missing and commit the results.
--- /go/src/k8s.io/kubernetes/Godeps/LICENSES	2016-05-25 08:56:54.604632383 -0700
+++ /go/src/k8s.io/kubernetes/_tmp/kube-godep-licenses.bPzSeq/Godeps/LICENSES	2016-05-25 09:11:39.829456100 -0700
@@ -64132,6 +64132,10 @@ SOFTWARE.
 
 
 ================================================================================
+= vendor/github.com/xanzy/go-cloudstack/cloudstack licensed under: =
+
+
+================================================================================
 = vendor/github.com/xiang90/probing licensed under: =
 
 The MIT License (MIT)
Removing workspace: /go/src/k8s.io/kubernetes/_tmp/kube-godep-licenses.bPzSeq
FAILED   ./hack/../hack/verify-godep-licenses.sh	19s
```
```
Verifying ./hack/../hack/verify-godeps.sh
Checking for 'Godeps/' changes against 'origin/master'
Starting to download all kubernetes godeps. This takes a while
Download finished
Your Godeps.json is different:
--- /go/src/k8s.io/kubernetes/Godeps/Godeps.json	2016-05-25 08:56:54.592632372 -0700
+++ /tmp/gopath.ZtYfxR/src/k8s.io/kubernetes/Godeps/Godeps.json	2016-05-25 09:16:58.633756823 -0700
@@ -1938,6 +1939,11 @@
 			"Rev": "9051bd6b44125d9472e0c148b5965692ab283d4a"
 		},
 		{
+			"ImportPath": "github.com/xanzy/go-cloudstack/cloudstack",
+			"Comment": "2.0.0-4-g9811b0f",
+			"Rev": "9811b0fde6c3d16da1b7f66bdcab47a3464e7d80"
+		},
+		{
 			"ImportPath": "github.com/xiang90/probing",
 			"Rev": "6a0cc1ae81b4cc11db5e491e030e4b98fba79c19"
 		},
@@ -2132,10 +2143,6 @@
 			"ImportPath": "k8s.io/heapster/metrics/apis/metrics/v1alpha1",
 			"Comment": "v1.1.0-beta2",
 			"Rev": "9cb18ac0ceb193eb530a1fe339355c94ea454d85"
-		},
-		{
-			"ImportPath": "github.com/xanzy/go-cloudstack/cloudstack",
-			"Rev": "9811b0fde6c3d16da1b7f66bdcab47a3464e7d80"
 		}
 	]
 }
Removing /tmp/gopath.ZtYfxR
FAILED   ./hack/../hack/verify-godeps.sh	408s
```
```
Verifying ./hack/../hack/verify-gofmt.sh
!!! 'gofmt -s' needs to be run on the following files: 
./pkg/cloudprovider/providers/providers.go
./pkg/cloudprovider/providers/cloudstack/cloudstack.go
FAILED   ./hack/../hack/verify-gofmt.sh	12s
```
```
ERROR: Step â€˜Publish JUnit test result reportâ€™ failed: No test report files were found. Configuration error?
```

File: pkg/cloudprovider/providers/cloudstack/MAINTAINERS.md (hunk 1 above)
Comment (runseb) [inline pkg/cloudprovider/providers/cloudstack/MAINTAINERS.md]: @lavalamp only the AWS and vsphere drivers actually have an OWNERS file.

**Comment by ngtuna** [comment]
@runseb @lavalamp Update progress: 

I made changes in local repo to fix the integration test. The problem relates to changes in Godeps dependencies. Before updating this PR I should run all verification scripts locally first. However right now I can not run all these scripts due to this issue: https://github.com/kubernetes/kubernetes/issues/25979.

**Comment by runseb** [comment]
@lavalamp @bgrant0607 hi guys, how can we move forward with this, to make sure it gets in a future release ? or potentially even 1.3 if there is still room and time ?

**Comment by lavalamp** [comment]
Hi @runseb we understand it's a pain when we sit on things, but please give us more time-- we are working hard on 1.3 right now. It will probably be at least a week before I have a chance to look at this. We won't forget it.

**Comment by runseb** [comment]
@lavalamp thanks, no problem, just want to make sure we are not missing anything.

**Comment by bgrant0607** [comment]
@runseb To clarify (in case it wasn't obvious): This can't make it into 1.3. 

@ngtuna Do you plan to set up continuous testing for this provider?

**Comment by ngtuna** [comment]
Hi @bgrant0607. Can you give me a hint how to set it up ?

**Comment by bgrant0607** [comment]
@ngtuna Do you have machines  you could run the tests on, or would you need resources?

@fejta Are there instructions for how to set up e2e for another cloud provider?

**Comment by ngtuna** [comment]
@bgrant0607 Do you mean this test? https://github.com/kubernetes/kubernetes/blob/master/docs/devel/e2e-tests.md

**Comment by fejta** [comment]
@bgrant0607 @ngtuna you want to run tests against another provider continuously? People seem to be using jenkins to do this, then copying the junit xml logs up to GCS and then letting https://kubernetes.slack.com/messages/sig-testing/details/ knows about it. There is not yet a canonical way of doing this jenkins setup. Please ask sig-testing for assitance. You can also try @ixdy

**Comment by mikedanese** [comment]
@k8s-bot ok to test

**Comment by runseb** [comment]
@ngtuna can you add an OWNERS file .thx

**Comment by ngtuna** [comment]
Yes. I already added OWNERS file but need to do the rebase again.

2016-07-12 14:04 GMT+07:00 runseb <notifications@github.com>:

**Comment by ngtuna** [comment]
Any hint on Jenkins verification failed ? I ran `./hack/../hack/verify-boilerplate.sh` locally without any issue.

cc/ @lavalamp @bgrant0607 @fejta @mikedanese @runseb

**Comment by runseb** [comment]
Not sure if we need to create an entry in https://github.com/kubernetes/features
Or which SIG we should talk about this

@aronchick ?

**Comment by mikedanese** [comment]
@k8s-bot ok to test, issue: #IGNORE

File: pkg/cloudprovider/providers/cloudstack/cloudstack.go [hunk 2]
```diff
@@ -0,0 +1,493 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
```
Comment (mikedanese) [inline pkg/cloudprovider/providers/cloudstack/cloudstack.go]: 2016

**Comment by mikedanese** [comment]
Can you put code changes in one commit and godep changes in another?

**Comment by ngtuna** [comment]
@mikedanese okay will do this today or tomorrow.

**Comment by svanharmelen** [comment]
@ngtuna please check your mail before doing so, as you might end up doing stuff double...

**Comment by ngtuna** [comment]
@svanharmelen Ah... I see your mail.

**Pull request closed without merging**